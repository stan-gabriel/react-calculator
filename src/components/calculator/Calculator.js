import React from 'react';
import { connect } from 'react-redux';
import Display from '../display/Display'
import Button from '../key-button/KeyButton';
import { rows, CalculatorOperations } from '../../utils/constants';
import * as actions from '../../reducers/actions';
import { CHANGE_SIGN, CLEAR_ALL, OPERATION, CALCULATE_RESULT, CALCULATE_PERCENTAGE } from '../../reducers/types';
import './Calculator.scss';


class Calculator extends React.Component {
	static defaultProps = {
		operation: '',
		addFirstValue: () => {
		}, addOperationType: () => {
		}, addSecondValue: () => {
		}, changeSign: () => {
		}, clearAll: () => {
		}, calculateResult: () => {
		}, calculatePercentage: () => {
		}
	};

	componentDidMount() {
		document.addEventListener('keydown', this.handleKeyDown)
	}

	componentWillUnmount() {
		document.removeEventListener('keydown', this.handleKeyDown)
	}

	performAction = (buttonData) => event => {
		const {operation, addFirstValue, addOperationType, addSecondValue, changeSign, clearAll, calculateResult, calculatePercentage} = this.props;

		if (buttonData.meaning === CHANGE_SIGN) {
			return changeSign();
		}
		if (buttonData.meaning === CLEAR_ALL) {
			return clearAll();
		}
		if (buttonData.meaning === OPERATION) {
			return addOperationType(buttonData.value, buttonData.action);
		}
		if (buttonData.meaning === CALCULATE_RESULT) {
			return calculateResult()
		}
		if (buttonData.meaning === CALCULATE_PERCENTAGE) {
			return calculatePercentage()
		}
		if (!operation) {
			addFirstValue(buttonData.value)
		} else {
			addSecondValue(buttonData.value);
		}
	};

	handleKeyDown = event => {
		const {operation, addFirstValue, addOperationType, addSecondValue, clearAll, calculateResult, calculatePercentage} = this.props;
		let {key} = event;

		if (key === 'Enter')
			calculateResult();
		if ((/\d/).test(key)) {
			// event.preventDefault();
			if (!operation) {
				addFirstValue(key)
			} else {
				addSecondValue(key);
			}
		} else if (key in CalculatorOperations) {
			// event.preventDefault();
			addOperationType(key, CalculatorOperations[key]);
		} else if (key === '.') {
			// event.preventDefault();
			if (!operation) {
				addFirstValue(key)
			} else {
				addSecondValue(key);
			}
		} else if (key === '%') {
			// event.preventDefault();
			calculatePercentage()
		} else if (key === 'Backspace') {
			// event.preventDefault();
			clearAll();
			// TODO implement 'clear last char'
		}
	};

	renderButton = (buttonData, index) => {
		return <Button key={index} colour={buttonData.colour} value={buttonData.value}
		               onClick={this.performAction(buttonData)}/>
	};

	renderRow = (row, index) => {
		return (
			<div className='buttons-wrapper' key={index}>
				{row.map(this.renderButton)}
			</div>
		)
	};

	render() {
		const {displayValue} = this.props;
		return (
			<div className="Calculator">
				<Display result={displayValue}/>

				{rows.map(this.renderRow)}
			</div>
		);
	}
}

export const PureCalculator = Calculator;

function mapStateToProps(state) {
	return {
		firstValue: state.app.firstValue,
		secondValue: state.app.secondValue,
		displayValue: state.app.displayValue,
		operation: state.app.operation,
	}
}

function mapDispatchToProps(dispatch) {
	return {
		addFirstValue: (firstValue) => dispatch(actions.addFirstValue(firstValue)),
		addSecondValue: (secondValue) => dispatch(actions.addSecondValue(secondValue)),
		addOperationType: (operationType, callback) => dispatch(actions.addOperationType(operationType, callback)),
		clearAll: () => dispatch(actions.clearAll()),
		changeSign: () => dispatch(actions.changeSign()),
		calculateResult: () => dispatch(actions.calculateResult()),
		calculatePercentage: () => dispatch(actions.calculatePercentage())
	}
}

export default connect(mapStateToProps, mapDispatchToProps)(Calculator);
