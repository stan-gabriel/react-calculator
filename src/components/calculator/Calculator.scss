@import '../../scss/variables';

.Calculator {
  display: flex;
  flex-direction: column;
  //height: 20rem;
  //width: 17rem;
  text-align: center;
  margin: 5rem;
}

.buttons-wrapper {
  display: flex;
}
