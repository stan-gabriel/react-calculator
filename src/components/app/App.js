import React from 'react';
import { Provider } from 'react-redux';
import { createStore, combineReducers } from 'redux';
import logo from '../../assets/logo.svg';
import './App.scss';
import '../calculator/Calculator'
import Calculator from "../calculator/Calculator";
import app from '../../reducers/app';

const store = createStore(combineReducers({
	app
}));

function App() {
	return (
		<Provider store={store}>
			<div className="App">

				<header className="App-header">
					<img src={logo} className="App-logo" alt="logo"/>
				</header>

				<Calculator />

			</div>
		</Provider>
	);
}

export default App;
