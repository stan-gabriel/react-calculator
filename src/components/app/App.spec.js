import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import Calculator from '../calculator/Calculator';

describe('App component', () => {
	let wrapper;
	beforeEach(() => (wrapper = shallow(<App />)));

	it('should match the snapshot', () => expect(wrapper).toMatchSnapshot());

	it('should render a <Provider />', () => {
		expect(wrapper.find('div').length).toEqual(1);
	});

	it('should render a <div />', () => {
		expect(wrapper.find('div').length).toEqual(1);
	});

	it('should render a <header />', () => {
		expect(wrapper.find('header').length).toEqual(1);
	});

	it('should render the Logo', () => {
		expect(wrapper.find('.App-logo').length).toEqual(1);
	});

	it('should render the Calculator Component', () => {
		expect(wrapper.containsMatchingElement(<Calculator />)).toEqual(true);
	});
});
