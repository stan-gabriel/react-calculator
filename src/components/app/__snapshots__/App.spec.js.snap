// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should match the snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <div
    className="App"
  >
    <header
      className="App-header"
    >
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
    </header>
    <Connect(Calculator) />
  </div>
</Provider>
`;
